### ISSUE TITLE
Minimize output parameter using neural network

### ISSUE THREAD

**Issue by samuelbertolini** [mlpack/mlpack#3087]
Dear all, sorry for the simple question, but I am having difficult to find a solution, and maybe this issue appeared here before. But the issue is:
I have the input parameters (Xi) that gives me outputs as a function of the parameters ( F[Xi] ). I would like to find the input parameters that minimize the function as much as possible. I was planning to use a neural network method in c++.

What are the most recommended methods for this? Can I do this by FNN? Is it more recommended to use GAN?

Thanks in advance!

**Comment by zoq** [comment]
Hello @samuelbertolini can you tell us more about your data, since you mentioned GAN, it's an image dataset, also is `F(x)` given?

**Comment by shrit** [comment]
I understood it as `x* = argmin f(x)`

**Comment by samuelbertolini** [comment]
What I have is a force field that I need to train. X are the parameters of the force field and F(x) is the error, where F(x) = Edft - Eff. Thus, I need to obtain x* = argmin(x). I though two way: FNN using random X and getting min F(x); or using GAN, entering a fake F(x) as input to obtain x*.
Is it clear now?

**Comment by shrit** [comment]
Yes, and you do not need to use complex methods such as FFN or GAN, if your function is differentiable just use an SGD from the ensmallen library.

**Comment by eddelbuettel** [comment]
And if it is not differentiable, there are other global (stochastic) search algorithms that can help (maybe at the price of slow progress -- the No Free Lunch Thereom (TM) still hangs around).

**Comment by samuelbertolini** [comment]
The problem is really not trivial. Stochastic gradient descent is already in use to minimize and to generate training data. The idea is to use ANN after SGD, because SGD is not enough. I just need to discover how to minimize the output.

**Comment by shrit** [comment]
ANN is not a function minimizer, they are function approximation methods, they will approximate the function and thus help you discover the function if you do not know it.

**Comment by shrit** [comment]
if you know `f(x)` then ANN is not usable here. But I think I still do not get the idea.
How ANN is going to help you?

**Comment by samuelbertolini** [comment]
Look this paper, https://www.sciencedirect.com/science/article/abs/pii/S092702562030598X
This gives the whole idea. I just need to do a similar implementation.

**Comment by shrit** [comment]
@samuelbertolini I am currently busy and I can not help you with the paper. However, mlpack has an example repository that has a set of neural networks examples, maybe GAN (I am not sure). https://github.com/mlpack/examples

**Comment by samuelbertolini** [comment]
No worries @shrit. I just showed the paper to show that ANN can be applied to find minimal function. I will take a look in the examples, thanks for the help!

**Comment by zoq** [comment]
In this case the paper uses a genetic algorithm (GA), that evolves the function (network architecture) over time that is used to optimize the function. Maybe you talked about GA from the beginning and I just got confused because you mentioned GAN at some point which for me is a generative adversarial network. Anyway ensmallen (https://github.com/mlpack/ensmallen) implements a bunch of evolution based methods (CMAES, CNE, DE) that you can test out as a starting point. That said there is also an open PR that implements NEAT (https://github.com/mlpack/mlpack/pull/1105) which worked on most of the tasks we tested. In addition some time ago I started working on https://arxiv.org/abs/1611.03824 which is basically a neural network learned optimizer.

**Comment by samuelbertolini** [comment]
Great @zoq . It is what I needed. Thanks a lot.
The paper used GA + FNN. I just though that it could be GA + GAN.
Anyway, this is really helpful.

**Comment by zoq** [comment]
> So, the evolution methods (CMAES, CNE, DE) were implemented inside a FNN?

Not exactly, similar to NEAT the methods implement a evolution strategy (mutation, offspring) to evolve the architecture, which could be interpreter as an FFN.

**Comment by samuelbertolini** [comment]
In the paper that I sent, if I am not wrong, they used FNN, right? How could they obtain minimization with FNN?
By the way, I will work to understand how to implement NEAT.
Thanks!

**Comment by samuelbertolini** [comment]
Sorry to bring up this again, but does anyone has a example for NEAT that I can follow and adapt?

**Comment by zoq** [comment]
Just any example or a more specific one?

**Comment by samuelbertolini** [comment]
More examples, the better it is. But one with any optimization would be great!

**Comment by samuelbertolini** [comment]
Sorry for asking again. Could you find any example for NEAT?

**Comment by zoq** [comment]
So first thing you need to do is to checkout https://github.com/mlpack/mlpack/pull/19 and build the `NEAT` branch. If you need help with that let me know.

The branch comes with a tutorial how to use the NEAT method: https://github.com/mlpack/mlpack/pull/1908/files#diff-e5644db8b584b1ca5285f29f479cab8e7e1d8d1429eda4c57f092060740743df and also implements some test cases https://github.com/mlpack/mlpack/pull/1908/files#diff-28bf203b8411d298e0afb2d4055dac5a628119f946e8b9c5eab5224d273cb139 that can serve as examples.

I can provide some more details for the test cases, if anything is unclear.

**Comment by samuelbertolini** [comment]
Thanks a lot @zoq. I hope that from now I can build my NEAT soon
